packages/vite/src/node/server/index.ts

```diff
@@ -480,33 +477,17 @@ export async function _createServer(
 
   const environments: Record<string, DevEnvironment> = {}
 
-  const client_createEnvironment =
-    config.environments.client?.dev?.createEnvironment ??
-    ((name: string, config: ResolvedConfig) =>
-      new DevEnvironment(name, config, { hot: ws, watcher }))
-
-  environments.client = await client_createEnvironment('client', config)
-
-  const ssr_createEnvironment =
-    config.environments.ssr?.dev?.createEnvironment ??
-    ((name: string, config: ResolvedConfig) =>
-      createNodeDevEnvironment(name, config, { hot: ssrHotChannel, watcher }))
-
-  environments.ssr = await ssr_createEnvironment('ssr', config)
-
-  for (const [name, EnvironmentOptions] of Object.entries(
+  for (const [name, environmentOptions] of Object.entries(
     config.environments,
   )) {
-    // TODO: move client and ssr inside the loop?
-    if (name !== 'client' && name !== 'ssr') {
-      const createEnvironment =
-        EnvironmentOptions.dev?.createEnvironment ??
-        ((name: string, config: ResolvedConfig) =>
-          new DevEnvironment(name, config, {
-            hot: ws, // TODO: what should we use here?
-          }))
-      environments[name] = await createEnvironment(name, config)
-    }
+    environments[name] = await environmentOptions.dev.createEnvironment(
+      name,
+      config,
+      {
+        hot: ws,
+        watcher,
+      },
+    )
```

Comment:
This is the main change in the PR. Not that before the PR, we were not passing `hot` and `watcher` to the Client and SSR environments when there was a custom `createEnvironment`. And we only pass `hot` in other environments, and only when they didn't have a custom `createEnvironment`. I wonder how this was affecting current experiments.

We're now passing `hot: ws` to every environment that isn't the `ssr` environment without a custom `createEnvironment`. @sheremet-va does this sounds ok? There is a TODO here that I'm removing.

Reply:
I didn't know there is a `watcher` property in `DevEnvironment`. It looks like it was only used for default client/ssr environments, but how does it affect the functionality?

Currently, for custom environment, I had only manually instantiate `DevEnvironment` with custom `{ hot }` like this https://github.com/hi-ogawa/vite-environment-examples/blob/0128d3b6bbe25263a57497fb7bb6dcae1c53df4b/packages/workerd/src/plugin.ts#L188

```ts
  const devEnv = new WorkerdDevEnvironmentImpl(name, config, { hot });
```

If `setup: { hot, watcher }` is provided in `createEnvironment` factory, then is it expected that `watcher` is also passed to `WorkerdDevEnvironmentImpl(..., { hot, watcher })`?

Reply:
> does this sounds ok? There is a TODO here that I'm removing.

I don't like this change combined with removing `hot: false`. I liked the explicitness of it, and I am not sure why websocket connection is treated as the main HMR source

Reply:
We need the watcher in the environment during `transformRequest` that nows takes and environment instead of the whole server. It is only used to ensure a fallback file is watched, so it is really not much. But we should force people to pass the setup. This is a good point towards @sheremet-va push to have these non-nullable. We could also have `watcher: false` to explicitly opt-out, and that would force users to do in `createEnvironment`:
```js
WorkerdDevEnvironmentImpl(..., { ...setup, hot })
```

About passing `{ hot: ws }` to every environment, I did this because we were already doing it for every environment that wasn't the `client` or `ssr` (and didn't have a custom `createEnvironment` function).

Maybe `createEnvironment()` third parameter should be different than the third parameter of `DevEnvironment()`. It could be
```
createEnvironment(name, config, { watcher, ws }) // this is how it is called
```
And then 
```
new DevEnvironment(name, config, { hot, watcher }) // required
```
And the user will create an environment as @hi-ogawa suggested.
```js
WorkerdDevEnvironmentImpl(..., { hot, watcher })
```
The issue here is that it is very easy to forget to pass the watcher. I think this parameter may be better pass to the init function and not part of the creation. So we could still pass the `watcher` to createEnvironment in case the user wants to use it, but it is not need in creation. It could just be:
```js
WorkerdDevEnvironmentImpl(..., { hot })
```
And then we will do as part of the creation flow `environment.init({ watcher })` (or directly `environment.watcher = watcher`) to make sure this is properly set.

We would also need a new `createClientDevEnvironment()` that sets `hot: ws` (as part of the config stage).

Just in case, trying to figure this myself too, this is showing that the current creation code for environments isn't consistent.